### ISSUE TITLE
differentiability/run inside tf.function decorated function

### ISSUE THREAD

**Issue by jonas-eschle** [N3PDF/vegasflow#76]
AFAIK, the integration cannot be performed within an actual graph. It is also not differentiable as it returns a float.

Both would be quite good to see in the library (such as in torchquad).

To give an idea, using the example from the readme:

<details>
<summary>Details</summary>

from vegasflow import vegas_wrapper
import tensorflow as tf

def integrand(x, **kwargs):
    """ Function:
       x_{1} * x_{2} ... * x_{n}
       x: array of dimension (events, n)
    """
    return tf.reduce_prod(x, axis=1)

dimensions = 8
iterations = 5
events_per_iteration = int(1e5)
vegas_wrapper(integrand, dimensions, iterations, events_per_iteration, compilable=True)

</details>

neither 
```
jitted = tf.function(vegas_wrapper)
vegas_wrapper(integrand, dimensions, iterations, events_per_iteration, compilable=True)
```
nor (deactivating the jit with `tf.config.execute_functions_eagerly()`)
```
with tf.GradientTape() as tape:
    y = vegas_wrapper(integrand, dimensions, iterations, events_per_iteration, compilable=True)[0]
tape.gradient(a, y)
```

I think the point is that vegasflow now considers itself as the highest level library. While this can be for some cases, integration can also be used *as part of a function* (say to integrate out a few dimensions).

What do you think about this? Or is there a way to achieve this already?

**Comment by scarlehoff** [comment]
The wrapper is high level so will never be differentiable. The reason for that is that there is high-level stuff going on basically to do with device/cluster management.

In order to differentiate the integration with the current form of the library you would need to create an instance of the integrator and probably call functions that right now are "hidden" (python hidden, with a `_`)

But this would be a nice addition. I'll have a look at how to expose something that (might) work.

Then there's the issue of being inside a `tf.function`, the wrapper is probably not a good idea but rather use https://vegasflow.readthedocs.io/en/latest/how_to.html#integrating-with-vegasflow
But I would not do `tf.function` but rather https://www.tensorflow.org/api_docs/python/tf/py_function

Now, I guess what you are actually interested in is using it inside a function that you can automaticallt differentiate. So there should be a differentiable wrapper-interface exposed by Vegasflow
This is something I will be working on as soon as I find the time because I think it would actually be quite useful. Thanks for the idea ^__^

**Comment by scarlehoff** [comment]
Actually, thinking about this, there's an "easy" way to get this behavior (I still think that an actual interface makes sense).

Since in this scenario I guess one is interested in the integration itself and not so much in the grid-refining process of Vegas, the derivative one would want is that of the latest iteration (so, only the result of the integration needs to be decorated). Then one can call directly `run_event` (instead of `run_iteration`), like this:

```python
from vegasflow import VegasFlow, float_me
import tensorflow as tf

dims = 4
n_calls = int(1e4)
vegas_instance = VegasFlow(dims, n_calls, verbose=False)
def example_integrand(x, **kwargs):
    y = 0.0
    for d in range(dims):
        y += x[:,d]
    return y

vegas_instance.compile(example_integrand)

# Now we run a few iterations to train the grid, but we can bin them
_ = vegas_instance.run_integration(3)

@tf.function
def some_complicated_function(x):
    integration_result, error, _ = vegas_instance.run_event()
    return x*integration_result

my_x = float_me(4.0)
result = some_complicated_function(my_x)

with tf.GradientTape() as tape:
    y = some_complicated_function(my_x)
tape.gradient(my_x, y)
```

But what if the integrand can change? Now you would need access to some internal functions

```python
from vegasflow import VegasFlow, float_me
import tensorflow as tf

dims = 4
n_calls = int(1e4)
vegas_instance = VegasFlow(dims, n_calls, verbose=False)

@tf.function
def some_complicated_function(x):
    
    def example_integrand(z, **kwargs):
        y = 0.0
        for d in range(dims):
            y += z[:,d] + x
        return y

    integration_result, error, _ = vegas_instance._run_event(example_integrand, n_calls)
    return integration_result

my_x = float_me(4.0)
result = some_complicated_function(my_x)

with tf.GradientTape() as tape:
    y = some_complicated_function(my_x)
tape.gradient(my_x, y)
```

In this second case I haven't refined the Vegas grid, the best thing is to do that inside a `py_function` environment (so that every time the integrand changes -for instance inside a NN-  the grid is re-refined).

In summary, it can already be done but 1) it is very much undocumented, 2) it would be nice to have a easy to use interface (maybe a wrapper that produces already a function with a reference to the right gradient).

In any case, thanks for the report, this is a way of using the code that we knew existed but we hadn't encountered yet :)

**Comment by jonas-eschle** [comment]
Thanks a lot for the thoughts, this is indeed what I had in mind. And I agree, the grid-refining process should be "outside", either run before or in a `py_function` like thing.

One possibility is to be able to inject the refinement parameterized (such as using the TFPs [AutoCompositeTensor](https://www.tensorflow.org/probability/api_docs/python/tfp/experimental/AutoCompositeTensor)) but that may gets ahead of things :)

But yes, an interface which is "optimized" for this would indeed be useful. Let me know if there is something I can do

**Comment by scarlehoff** [comment]
An interface for this is given in #79. By calling `make_differentiable` one gets a reference to a function that upon calling returns the integration result.
